### ISSUE TITLE
Add English keyboard

### ISSUE THREAD

**Issue by andrewtavis** [scribe-org/Scribe-iOS#7]
### Terms

- [X] I have searched [open new keyboard issues](https://github.com/scribe-org/Scribe-iOS/issues?q=is%3Aopen+is%3Aissue+label%3A%22new+keyboard%22)
- [X] I agree to follow Scribe-iOS' [Code of Conduct](https://github.com/scribe-org/Scribe-iOS/blob/main/.github/CODE_OF_CONDUCT.md)
- [X] This issue is about one language, and I've changed the title to reflect this

### What's needed

Adding an English keyboard to Scribe is as of now not a priority, but it will eventually be added, and could be worked on now if there's interest. There is as of now a good amount of data on [Wikidata](https://www.wikidata.org/) as can be seen in the [queries](https://github.com/scribe-org/Scribe-iOS/blob/main/Data/English/). The English keyboard will be more of a focus for Scribe when the `Translate` feature is more developed [(see project)](https://github.com/scribe-org/Scribe-iOS/projects/1), and further when the `Synonym` feature is added.

What's needed:
- [x] English language keyboard values added to [InterfaceVariables.swift](https://github.com/scribe-org/Scribe-iOS/blob/main/Keyboards/KeyboardsBase/InterfaceVariables.swift)
- [x] Fine tuning the WDQS queries found in [Scribe-iOS/Data/English](https://github.com/scribe-org/Scribe-iOS/tree/main/Data/English)
- [x] Writing formatting Python scripts and adding them to [Scribe-iOS/Data/English](https://github.com/scribe-org/Scribe-iOS/tree/main/Data/English)
- [ ] Add English keyboard specific logic to [KeyboardViewController.swift](https://github.com/scribe-org/Scribe-iOS/blob/main/Keyboards/KeyboardsBase/KeyboardViewController.swift)
- [x] Setting up alternate key functions if there are keys that have them that aren't already implemented
- [ ] Adding a English directory to [Scribe-iOS/Keyboards/LanguageKeyboards](https://github.com/scribe-org/Scribe-iOS/tree/main/Keyboards/LanguageKeyboards) that includes all the formatted JSON data and an `EnglishKeyboardViewController.swift` file
- [ ] Adding all the formatted SQLite data to a `Data` directory in the new directory and updating the data table in the readme
- [ ] Add English keyboard specific logic to [CommandVariables.swift](https://github.com/scribe-org/Scribe-iOS/blob/main/Keyboards/KeyboardsBase/CommandVariables.swift)

**Comment by muescha** [comment]
i would vote for an english keyboard - i think the most students are learning english and this keyboard would be helpful

**Comment by andrewtavis** [comment]
Thanks for your feedback, @muescha!

Yes I think that an English keyboard would help for irregular verbs. Issue is that we’re currently only translating from English to other languages. This issue needs to wait until at least #16 is finished as a user would need to be able to select a language to translate from, and then we could add the translation JSONs. 

Any further feedback or support you’d like to provide would be very welcome! Again, thanks for reaching out! 😊

**Comment by andrewtavis** [comment]
- Issue is currently blocked by #16 as people need to be able to select which language they would translate from into English

**Comment by wkyoshida** [comment]
As mentioned in this [discussion](https://github.com/scribe-org/Scribe-Data/issues/14#issuecomment-1299537864), I would be curious to hear thoughts on where the English keyboard now fits into the roadmap. I feel that the need is definitely there :pray:

**Comment by andrewtavis** [comment]
Glad to revisit this, @wkyoshida 😊

Big thing as I mentioned above is that we need to be able to choose which language to translate from. Listening to the interest here, I think it makes sense that we do something along the lines of translations being an option for any language pair that we can find a viable [🤗 Hugging Face](https://huggingface.co/) model for. This would mean that we can do English to any of the seven languages as we do now, and then any of the seven languages to English. A translation JSON is roughly 2MB, so I don't think that this is preventative for adding any individual language pair. It's not a guarantee that there will be a valid translation model for each language pair, but even if we use a general one it should be ok. We can check the results, and worst comes to worst one translation pair isn't included.

Ultimately we'd need to edit the translation scripts in Scribe-Data to basically generate the translation files where the given language is no longer simply the target language from English, but that each of the seven models (we'd have eight total languages then) is ran and we get the resultant JSONs stored. We'd need to change the architecture here to check what the user has selected, as a base language, with allowing for system language selection being a great baseline that they could then choose from. Defaults for if system language isn't offered would be English for each language, and I'd argue Spanish should be the default base for English.

I think I can update the roadmap based on the above. The English keyboard would definitely come before data downloads :) Question is whether we think that this would qualify for v3.0, which to me it kind of does 😊

**Comment by andrewtavis** [comment]
Another thing we're hoping to do at some point, @wkyoshida, is to do a launch on [Product Hunt](https://www.producthunt.com/). I think that as soon as we have the above translation pairs in place we'd be ready to make a splash 😊

**Comment by wkyoshida** [comment]
Gotchu! Thanks - I think that what you shared makes sense, @andrewtavis! :+1: 
I did notice as well the new issue created for the translation data, https://github.com/scribe-org/Scribe-Data/issues/23.

> Another thing we're hoping to do at some point, @wkyoshida, is to do a launch on [Product Hunt](https://www.producthunt.com/). I think that as soon as we have the above translation pairs in place we'd be ready to make a splash :blush:

Oh - I see! That would be pretty exciting :raised_hands: I just noticed it also has been in the outreach plans, https://github.com/scribe-org/Scribe-iOS/issues/181

**Comment by ayushanand308** [comment]
@andrewtavis I want to work on this issue and I am new to this project. Can you share some resources on how to get started so that I can work on this issue?

**Comment by andrewtavis** [comment]
96f32ab adds the directory for the eventual English keyboard :)

**Comment by wkyoshida** [comment]
I can't remember if this is something that we even considered, but could it be possible if we just add the English keyboard, although without a translation option for it? 

This would make it the only keyboard without the feature unfortunately, but I'm wondering if the need/desire for English could justify this. The other keyboard features would still be beneficial to users I think. I'm just thinking about the issues that are blocking a complete English keyboard and https://github.com/scribe-org/Scribe-Data/issues/23 seems it could potentially take some time to get completed. My guess, as well, is that the user desire for English is higher than that for Bokmål or Danish.

Something I was wondering recently. Let me know :shrug:

**Comment by andrewtavis** [comment]
I totally agree that there's more need for English than there is for any other language :) I definitely do want to include a translate feature with it as well, but then I think that we can actually start on this pretty soon. My assumption is that #96 will be done soonish so long as the SQLite links are intuitive, and from there it's a data update away — one that will need me to check the process, but it'll be all of a couple hours max to make sure it's back to normal after #96. My hope is by the end of the week 😊

Something we could start on now is the WDQS queries and the Python formatting scripts above?

**Comment by wkyoshida** [comment]
> Something we could start on now is the WDQS queries and the Python formatting scripts above? Would you have an interest in looking at those, @wkyoshida?

I could give that a go :+1: unless though if we would like to loop it under as one of the items for @SaurabhJamadagni to take on as part of GSoC. We can talk through a plan for this on our call!

**Comment by andrewtavis** [comment]
I’d say so long as @SaurabhJamadagni does a keyboard it’s all good :) Doesn’t have to be this one, and we have #135 and #133 that would also be fine 😊 Let’s discuss!

**Comment by andrewtavis** [comment]
Closed by #484 :) Sooo looking forward to getting the translation process set and then getting this keyboard into the App Store! Thanks for all the hard work, @Jag-Marcel! 😊
